tests/imputers/test_ModeImputer.py

```diff
@@ -34,70 +41,175 @@ class TestFit(WeightColumnFitMixinTests, GenericFitTests):
     def setup_class(cls):
         cls.transformer_name = "ModeImputer"
 
-    def test_learnt_values(self):
+    @pytest.mark.parametrize(
+        "library",
+        ["pandas", "polars"],
+    )
+    def test_learnt_values(self, library):
         """Test that the impute values learnt during fit are expected."""
-        df = d.create_df_3()
+        df = d.create_df_6(library=library)
 
         x = ModeImputer(columns=["a", "b", "c"])
 
         x.fit(df)
 
-        ta.classes.test_object_attributes(
-            obj=x,
-            expected_attributes={
-                "impute_values_": {
-                    "a": df["a"].mode()[0],
-                    "b": df["b"].mode()[0],
-                    "c": df["c"].mode()[0],
-                },
-            },
-            msg="impute_values_ attribute",
-        )
+        expected_impute_values = {
+            "a": 2,
+            "b": "a",
+            "c": "f",
+        }
+
+        assert (
+            x.impute_values_ == expected_impute_values
+        ), f"impute_values_ attribute not expected, expected {expected_impute_values} but got {x.impute_values_}"
+
+    @pytest.mark.parametrize(
+        "library",
+        ["pandas", "polars"],
+    )
+    def test_learnt_values_tied(self, library):
+        """Test that the impute values learnt during fit are expected - when mode is tied."""
+        df = d.create_df_3(library=library)
+
+        x = ModeImputer(columns=["a"])
+
+        with pytest.warns(
+            UserWarning,
+            match="ModeImputer: The Mode of column a is tied, will sort in descending order and return first candidate",
+        ):
+            x.fit(df)
+
+        expected_impute_values = {
+            "a": 6,
+        }
 
-    def expected_df_nan():
-        return pd.DataFrame({"a": ["NaN", "NaN", "NaN"], "b": [None, None, None]})
+        assert (
+            x.impute_values_ == expected_impute_values
+        ), f"impute_values_ attribute not expected, expected {expected_impute_values} but got {x.impute_values_}"
 
     @pytest.mark.parametrize(
-        ("df", "expected"),
-        ta.pandas.row_by_row_params(
-            pd.DataFrame({"a": [np.nan, np.nan, np.nan], "b": [None, None, None]}),
-            expected_df_nan(),
+        "library",
+        ["pandas", "polars"],
+    )
+    def test_learnt_values_tied_weighted(self, library):
```

Comment:
S: should we maybe consider testing this behaviour with types other than int? e.g str?  Sort of thing where behaviour could maybe change with changes in dependencies and cause a real head scratcher.

Reply:
sorry this is probably a distant memory now, but finally got back to this https://github.com/lvgig/tubular/pull/345/commits/37771b14f87e9b3654461b8294e5db4327f31344

adding more type tests did expose some issues, so was worth doing, you'll see I had to make some improvements to dataframe_init_dispatch (and also added tests)

Reply:
thanks for adding this